libc/stdlib_bsearch.c

```diff
@@ -0,0 +1,320 @@
+		test_clean(names);
+	}
+}
+
+
+TEST(bsearch, find_node_not_in)
+{
+	struct node *rval, node;
+	char name[] = "nonPresentName";
+	node.string = name;
+	node.length = strlen(name);
+
+	/* Check for even number of elements */
+	test_prepare(names);
+
+	rval = (struct node *)bsearch((void *)&node, (void *)test_table,
+		(size_t)TABSIZE, sizeof(struct node), test_nodeCmp);
+	TEST_ASSERT_NULL(rval);
+
+	TEST_ASSERT_EQUAL_INT(1, test_checkTab(names));
+	test_clean(names);
+
+	/* Check for odd number of elements */
+	test_prepare(names);
+
+	rval = (struct node *)bsearch((void *)&node, (void *)test_table,
+		(size_t)TABSIZE - 1, sizeof(struct node), test_nodeCmp);
+	TEST_ASSERT_NULL(rval);
+
+	TEST_ASSERT_EQUAL_INT(1, test_checkTab(names));
+	test_clean(names);
+}
+
+
+TEST(bsearch, not_sorted)
+{
+	/* Elements are not sorted in this way that bsearch won't find searched element */
+	struct node *rval, node;
+	char name[] = "adam";
+	node.string = name;
+	node.length = strlen(name);
+
+	/* Check for even number of elements */
+	test_prepare(names_nsort);
+
+	rval = (struct node *)bsearch((void *)&node, (void *)test_table,
+		(size_t)TABSIZE, sizeof(struct node), test_nodeCmp);
+	TEST_ASSERT_NULL(rval);
+
+	TEST_ASSERT_EQUAL_INT(1, test_checkTab(names_nsort));
+	test_clean(names_nsort);
+
+	/* Check for odd number of elements */
+	test_prepare(names_nsort);
+
+	rval = (struct node *)bsearch((void *)&node, (void *)test_table,
+		(size_t)TABSIZE, sizeof(struct node), test_nodeCmp);
+	TEST_ASSERT_NULL(rval);
+
+	TEST_ASSERT_EQUAL_INT(1, test_checkTab(names_nsort));
+	test_clean(names_nsort);
+}
+
+TEST(bsearch, same_members)
+{
+	struct node *rval, node;
+	char name[] = "adam";
+	node.string = name;
+	node.length = strlen(name);
+
+	/* Check for even number of elements */
+	test_prepare(namesRepeat);
+
+	rval = (struct node *)bsearch((void *)&node, (void *)test_table,
+		(size_t)TABSIZE, sizeof(struct node), test_nodeCmp);
+	TEST_ASSERT_NOT_NULL(rval);
+	TEST_ASSERT_EQUAL_STRING(node.string, rval->string);
+	TEST_ASSERT_EQUAL_INT(node.length, rval->length);
+
+	TEST_ASSERT_EQUAL_INT(1, test_checkTab(namesRepeat));
+	test_clean(namesRepeat);
+
+	/* Check for odd number of elements */
+	test_prepare(namesRepeat);
+
+	rval = (struct node *)bsearch((void *)&node, (void *)test_table,
+		(size_t)TABSIZE - 1, sizeof(struct node), test_nodeCmp);
+	TEST_ASSERT_NOT_NULL(rval);
+	TEST_ASSERT_EQUAL_STRING(node.string, rval->string);
+	TEST_ASSERT_EQUAL_INT(node.length, rval->length);
+
+	TEST_ASSERT_EQUAL_INT(1, test_checkTab(namesRepeat));
+	test_clean(namesRepeat);
+}
+
+
+TEST(bsearch, special_node)
+{
+	struct node *rval;
+	struct node2 node2;
+	char name[] = "adam";
+	node2.string = name;
+	node2.length = strlen(name);
+
+	/* Check for even number of elements */
+	test_prepare(names);
+
+	rval = (struct node *)bsearch((void *)&node2, (void *)test_table,
+		(size_t)TABSIZE, sizeof(struct node), test_nodeCmp2);
+	TEST_ASSERT_NULL(rval);
```

Comment:
Please explain this case a little bit in a comment, the best would be o provide and example, so sth like: 

```C
/* the table content is {str1len1str2len2str3len3}, so we assert that elements are splitted properly and len1str2 won't be found */
```